worker/raft/raftclusterer/worker_test.go

```diff
@@ -0,0 +1,302 @@
+	s.Transport.Connect(transport1.LocalAddr(), transport1)
+	s.Transport.Connect(transport2.LocalAddr(), transport2)
+	transport1.Connect(transport2.LocalAddr(), transport2)
+	transport2.Connect(transport1.LocalAddr(), transport1)
+	machine1Address := string(transport1.LocalAddr())
+	machine2Address := string(transport2.LocalAddr())
+
+	raft1Observations := make(chan raft.Observation, 10)
+	raft1Observer := raft.NewObserver(raft1Observations, false, func(o *raft.Observation) bool {
+		_, ok := o.Data.(raft.LeaderObservation)
+		return ok
+	})
+	raft1.RegisterObserver(raft1Observer)
+	defer close(raft1Observations)
+	defer raft1.DeregisterObserver(raft1Observer)
+
+	newLeaderElected := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case _, ok := <-raft1Observations:
+				if !ok {
+					return
+				}
+			}
+			leaderAddr := raft1.Leader()
+			if leaderAddr != "" && leaderAddr != s.Transport.LocalAddr() {
+				close(newLeaderElected)
+				return
+			}
+		}
+	}()
+
+	// Here we simulate "ensure-ha": machine-0's address will
+	// be updated to a non-localhost address, and two new
+	// servers are added.
+	//
+	// We add machine-1 and machine-2, and change machine-0's
+	// address. The new machines should be added first, and
+	// then machine-0 should be *removed* from the configuration,
+	// under the expectation that one of the new servers will
+	// become leader and add it back to the configuration with
+	// its new address.
+	s.publishDetails(c, apiserver.Details{
+		Servers: map[string]apiserver.APIServer{
+			"0": {
+				ID:              "0",
+				InternalAddress: "testing.invalid:1234",
+			},
+			"1": {
+				ID:              "1",
+				InternalAddress: machine1Address,
+			},
+			"2": {
+				ID:              "2",
+				InternalAddress: machine2Address,
+			},
+		},
+	})
+	waitConfiguration(c, raft1, []raft.Server{{
+		ID:       "machine-1",
+		Address:  raft.ServerAddress(machine1Address),
+		Suffrage: raft.Voter,
+	}, {
+		ID:       "machine-2",
+		Address:  raft.ServerAddress(machine2Address),
+		Suffrage: raft.Voter,
+	}})
+
+	select {
+	case <-newLeaderElected:
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out waiting for leader observation")
+	}
+}
+
+func (s *WorkerSuite) publishDetails(c *gc.C, details apiserver.Details) {
+	received, err := s.hub.Publish(apiserver.DetailsTopic, details)
+	c.Assert(err, jc.ErrorIsNil)
+	select {
+	case <-received:
+	case <-time.After(coretesting.LongWait):
+		c.Fatal("timed out waiting for details to be received")
+	}
+}
+
+func waitConfiguration(c *gc.C, r *raft.Raft, expectedServers []raft.Server) {
+	var configuration raft.Configuration
+	for a := coretesting.LongAttempt.Start(); a.Next(); {
+		f := r.GetConfiguration()
+		c.Assert(f.Error(), jc.ErrorIsNil)
+		configuration = f.Configuration()
+		if reflect.DeepEqual(configuration.Servers, expectedServers) {
+			return
+		}
+	}
+	c.Assert(configuration.Servers, jc.SameContents, expectedServers)
```

Comment:
probably a gc.Commentf("waited %s and still did not see the expected configuration", coretesting....)

Reply:
done